crypto/crypto.go

```diff
@@ -0,0 +1,153 @@
+	// view 0 is always valid.
+	if view == 0 {
+		return consensus.NewTimeoutCert(nil, 0), nil
+	}
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		sigs = append(sigs, timeout.ViewSignature)
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.TimeoutCert{}, err
+	}
+	return consensus.NewTimeoutCert(sig, view), nil
+}
+
+// CreateAggregateQC creates an AggregateQC from the given timeout messages.
+func (c crypto) CreateAggregateQC(view consensus.View, timeouts []consensus.TimeoutMsg) (aggQC consensus.AggregateQC, err error) {
+	qcs := make(map[hotstuff.ID]consensus.QuorumCert)
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		if qc, ok := timeout.SyncInfo.QC(); ok {
+			qcs[timeout.ID] = qc
+		}
+		if timeout.MsgSignature != nil {
+			sigs = append(sigs, timeout.MsgSignature)
+		}
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.AggregateQC{}, err
+	}
+	return consensus.NewAggregateQC(qcs, sig, view), nil
+}
+
+// VerifyPartialCert verifies a single partial certificate.
+func (c crypto) VerifyPartialCert(cert consensus.PartialCert) bool {
+	block, ok := c.mods.BlockChain().Get(cert.BlockHash())
+	if !ok {
+		return false
+	}
+	return c.Verify(cert.Signature(), block.ToBytes())
+}
+
+// VerifyQuorumCert verifies a quorum certificate.
+func (c crypto) VerifyQuorumCert(qc consensus.QuorumCert) bool {
+	// genesis QC is always valid.
+	if qc.BlockHash() == consensus.GetGenesis().Hash() {
+		return true
+	}
+	if qc.Signature().Participants().Len() < c.mods.Configuration().QuorumSize() {
+		return false
+	}
+	block, ok := c.mods.BlockChain().Get(qc.BlockHash())
+	if !ok {
+		return false
+	}
+	return c.Verify(qc.Signature(), block.ToBytes())
+}
+
+// VerifyTimeoutCert verifies a timeout certificate.
+func (c crypto) VerifyTimeoutCert(tc consensus.TimeoutCert) bool {
+	// view 0 TC is always valid.
+	if tc.View() == 0 {
+		return true
+	}
+	if tc.Signature().Participants().Len() < c.mods.Configuration().QuorumSize() {
+		return false
+	}
+	return c.Verify(tc.Signature(), tc.View().ToBytes())
+}
+
+// VerifyAggregateQC verifies the AggregateQC and returns the highQC, if valid.
+func (c crypto) VerifyAggregateQC(aggQC consensus.AggregateQC) (highQC consensus.QuorumCert, ok bool) {
+	messages := make(map[hotstuff.ID][]byte)
+	for id, qc := range aggQC.QCs() {
+		if highQC.View() < qc.View() || highQC == (consensus.QuorumCert{}) {
+			highQC = qc
+		}
+		// reconstruct the TimeoutMsg to get the hash
+		messages[id] = consensus.TimeoutMsg{
+			ID:       id,
+			View:     aggQC.View(),
+			SyncInfo: consensus.NewSyncInfo().WithQC(qc),
+		}.ToBytes()
+	}
+	if aggQC.Sig().Participants().Len() < c.mods.Configuration().QuorumSize() {
+		return consensus.QuorumCert{}, false
+	}
+	ok = c.BatchVerify(aggQC.Sig(), messages)
```

Comment:
Move into if?

Reply:
Actually, this code (Lines 145-152) could be written as:
```go
if c.BatchVerify(aggQC.Sig(), messages) && c.VerifyQuorumCert(highQC) {
	return highQC, true
}
return consensus.QuorumCert{}, false
```
Perhaps adding a comment above the if to say that both must verify...